src/qibolab/instruments/simulator/backends/generic.py

```diff
@@ -0,0 +1,284 @@
+from collections import OrderedDict
+from typing import List, Optional, Union
+
+import numpy as np
+from IPython.display import Latex, display
+
+
+def dec_to_basis_string(
+    x: int, nlevels: list = [2], to_string: bool = True
+) -> Union[str, list]:
+    """Converts an integer to a generalized bitstring in the computation basis
+    of the full Hilbert space.
+
+    Args:
+        x (int): The integer (decimal number) to be converted.
+        nlevels (list): The list of nlevels to convert the decimal number to. Defaults to [2].
+        to_string (bool): Flag to return the result as a string. Defaults to True.
+
+    Returns:
+        str or list: Generalized bitstring of x. If to_string is True, the result is returned as a string, otherwise as a list of integers.
+    """
+
+    nqubits = len(nlevels)
+
+    output_list = []
+    y = x
+    for i in range(nqubits):
+        if i < nqubits - 1:
+            sub_dim = np.prod(nlevels[i + 1 :])
+            coeff = np.divmod(y, sub_dim)[0]
+            output_list.append(coeff)
+            y -= coeff * sub_dim
+        else:
+            coeff = np.divmod(y, 1)[0]
+            output_list.append(coeff)
```

Comment:
Sorry to post the full historic, but I just wanted to make you notice that in the way described above you don't even need the `else:` branch, since it is a special case of the `if:` one.
```suggestion
    output_list = []
    y = x
    
    subdims_ = np.mulitply.accumulate(nlevels)
    subdims = (subdims_[-1] / subdims_).astype(int)
    for sub_dim in subdims:
        coeff = np.divmod(y, sub_dim)[0]
        output_list.append(coeff)
        y -= coeff * sub_dim
```

Reply:
I actually suspect that even the rest of the `for` loop could be implemented just as idiomatic NumPy operations, avoiding the loop *tout court*. I just haven't tried to go all the way.

Reply:
I'm not sure I can do better within a justifiable amount of time so I will just implement your simplification first, thanks!

Reply:
Fixed in 4aa6551 and 7634859